internal/persistence/sql/migrations/templates/20210623162417_relationtuple.mysql.up.sql

```diff
@@ -0,0 +1,25 @@
+-- we can not fizz this migration because there is no check constraint support in fizz
+-- and it gets cumbersome with sqlite; having a working SQL version is actually way easier
+CREATE TABLE keto_relation_tuples
+(
+    shard_id                 char(36)    NOT NULL,
+    nid                      char(36)    NOT NULL,
+    namespace_id             INTEGER     NOT NULL,
```

Comment:
Wouldn't this compose the primary key also? And if so, would it make sense to assign it a UUID? An int might become difficult in multi-tenant scenarios where you have lots of namespaces in lots of nids

Reply:
Yes, but why add the `namespace_id` to the PK?
The namespace ID is an int because it is managed by the user (maybe add docs for that :sweat_smile:). It determines the mapping between the namespace name (used in the API) and the actual database entry. This allows easy renaming of the namespaces as renaming does not affect the database at all. It was like that already before, is there a reason to make this a breaking change? Would you then create the UUIDs internally? How do you then handle changes to the namespace?

Reply:
Yeah, the problem in Ory Cloud will be that we have e.g. 100.000 namespaces, and potentially several customers creating namespaces concurrently (specifically across regions) which would mean that we have to do a table lock in order to fetch the highest ID (basically the auto_increment issue) on potentially billions of elements (as we have all relations in here). So it would be better to have something unique here, IMO.

I don't know how that could be solved in the user config. Potentially with some additional table / mapping:

```
struct NamespaceMapping {
  ConfigKey int
  ID uuid.UUID
}
```

which would be used for config mapping.

Given that we probably need some config in the db at some point, it might make sense to have it a

```
struct NamespaceConfig {
  ID uuid.UUID
  ConfigID int // can be empty
}
```

Reply:
But the namespace ID only has to be unique within a network. Also it has to be provided by some outside source (config.yml). We only have to increment it for every project, which is a searchable scope. UUIDs on the other hand will be annoying to manage manually because it is hard to compare them and a multi step process to generate new ones manually.
Think of the namespace rather as a directory or bucket of tuples that helps you organize it. Also they will contain the subject set rewrites and effect of relations.

Reply:
I get it, you can define the namespace with an ID in your config file:

```
namespaces:
  - id: 1
    name: foo
```

but in Ory Keto it would translate into a new row in table `namespace_configs`:

```
INSERT INTO namespace_configs VALUES (id=newUUID, config_id=1)
```

So you could use UUIDs in the relation tuples, but don't have to lock the row with high isolation levels to find the largest ID in the search space, which would be a filter + order query on a insanely large distributed database table.

Reply:
I think I now got what you initially meant here

> I don't know how that could be solved in the user config. Potentially with some additional table / mapping:
> 
> ```
> struct NamespaceMapping {
>   ConfigKey int
>   ID uuid.UUID
> }
> ```
>
> which would be used for config mapping.

Sounds good, so this is only used when loading namespace configs from the config file, but not e.g. from a database right? When loading from the database you directly specify the UUIDs.